core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -576,7 +579,8 @@ public int compare(ShardRouting o1,
                     if ((indexCmp = o1.index().compareTo(o2.index())) == 0) {
                         return o1.getId() - o2.getId();
                     }
-                    return indexCmp;
+                    final int secondary = secondaryComparator.compare(o1, o2);
+                    return secondary == 0 ? indexCmp : secondary;
```

Comment:
This was a bit hard for me to read due to the order in which comparators are checked. Could it be rewritten in a more idiomatic way, ie.

```java
if (o1.isPrimary() != o2.isPrimary()) {
  return o1.isPrimary() ? -1 : 1;
}
final int secondaryCmp = secondaryComparator.compare(o1, o2);
if (secondaryCmp != 0) {
  return secondaryCmp;
}
final int indexCmp = o1.index().compareTo(o2.index()));
if (indexCmp != 0) {
  return indexCmp;
}
final int idCmp = o1.getId() - o2.getId();
if (idCmp != 0) {
  return idCmp;
}
```
It would be helpful at least for me to see more quickly in which order comparisons are performed.